polar-core/src/data_filtering.rs

```diff
@@ -1158,3 +1223,447 @@ mod test {
+                    args[1].value().as_string()?.to_string(),
+                ),
+                self.data_source(Self::path_disasm(&args[0])?),
+            )),
+            _ => err_unimplemented(format!("datum({})", t.to_polar())),
+        }
+    }
+
+    fn binary_op(mut self, args: &[Term], op: SelOp) -> PolarResult<Self> {
+        let (l, r) = (self.datum(&args[0])?, self.datum(&args[1])?);
+        self.constraints.push((l, op, r));
+        Ok(self)
+    }
+
+    fn source(&mut self, t: &Term) -> PolarResult<Rc<DataFilter>> {
+        use DataFilter::Source;
+        let path = Self::path_disasm(t)?;
+        match self.entities.get(&path) {
+            Some(src) => Ok(Rc::new(Source(src.clone()))),
+            _ => err_invalid(format!("No entity for {:?}", path)),
+        }
+    }
+
+    fn new(types: Types, parts: Vec<Operation>) -> PolarResult<Self> {
+        use Operator::*;
+        let (isas, othas): (Vec<_>, Vec<_>) = parts
+            .into_iter()
+            .partition(|x| matches!(x, Operation { operator: Isa, args } if args.len() == 2));
+        let mut entities = isas
+            .into_iter()
+            .map(|Operation { args, .. }| {
+                Ok((
+                    Self::path_disasm(&args[0])?,
+                    args[1].value().to_class_tag()?,
+                ))
+            })
+            .collect::<PolarResult<HashMap<Vec<String>, String>>>()?;
+        let source_tag = entities
+            .get(&vec!["_this".to_string()])
+            .unwrap()
+            .to_string();
+
+        if let Some(fields) = types.get(&source_tag) {
+            for (field, typ) in fields {
+                if let Type::Relation {
+                    other_class_tag, ..
+                } = typ
+                {
+                    entities.insert(
+                        vec!["_this".to_string(), field.clone()],
+                        other_class_tag.clone(),
+                    );
+                }
+            }
+        }
+
+        let source = Rc::new(DataFilter::Source(source_tag));
+        let constraints = vec![];
+        let this = Self {
+            source,
+            types,
+            entities,
+            constraints,
+        };
+        othas
+            .into_iter()
+            .fold(Ok(this), |this, opn| this?.constrain(&opn))
+    }
+
+    fn constrain(self, part: &Operation) -> PolarResult<Self> {
+        use Operator::*;
+        match part.operator {
+            And => part.args.iter().fold(Ok(self), |this, c| {
+                this?.constrain(c.value().as_expression().unwrap())
+            }),
+            Isa if part.args.len() == 2 => {
+                Ok(self)
+            }
+            Eq | Unify | Assign if part.args.len() == 2 => {
+                self.binary_op(&part.args, SelOp::Eq)
+            }
+            In if part.args.len() == 2 => {
+                self.binary_op(&part.args, SelOp::In)
+            }
+            Neq if part.args.len() == 2 => {
```

Comment:
Do we need to check here that Isa has the right type? Also, this isa will only show up for sub fields right (otherwise would be pulled out in `new`)